src/control/functor/multivariate.lean

```diff
@@ -0,0 +1,770 @@
+/-
+Copyright (c) 2018 Jeremy Avigad. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Jeremy Avigad, Mario Carneiro, Simon Hudon
+-/
+import data.fin2
+import logic.function.basic
+import tactic.basic
+
+/-!
+
+Tuples of types, and their categorical structure.
+
+Features:
+
+`typevec n`   : n-tuples of types
+`α ⟹ β`      : n-tuples of maps
+`f ⊚ g`       : composition
+`mvfunctor n` : the type class of multivariate functors
+`f <$$> x`    : notation for map
+
+Also, support functions for operating with n-tuples of types, such as:
+
+`append1 α β`    : append type `β` to n-tuple `α` to obtain an (n+1)-tuple
+`drop α`         : drops the last element of an (n+1)-tuple
+`last α`         : returns the last element of an (n+1)-tuple
+`append_fun f g` : appends a function g to an n-tuple of functions
+`drop_fun f`     : drops the last function from an n+1-tuple
+`last_fun f`     : returns the last function of a tuple.
+
+Since e.g. `append1 α.drop α.last` is propositionally equal to `α` but not definitionally equal
+to it, we need support functions and lemmas to mediate between constructions.
+-/
+
+universes u v w
+
+
+/--
+n-tuples of sorts, as a category
+-/
+def ptypevec (n : ℕ) := fin2 n → Sort*
```

Comment:
Most of the definitions in this file are about `typevec` and not `ptypevec`.  Is the more general `ptypevec` actually used anywhere?

Reply:
I thought it was still useful, I'm looking through the datatype compiler and it seems to have disappeared. I can take it out and put it back if it turns out to be necessary

Reply:
Please!